src/ui/public/vis/vis_types/vislib_vis_type.js

```diff
@@ -17,25 +19,70 @@ export function VislibVisTypeProvider(Private) {
     constructor(el, vis) {
       this.el = el;
       this.vis = vis;
+
+      this.container = document.createElement('div');
+      this.container.className = 'vislib-container';
+      this.el.appendChild(this.container);
+
+      this.chartEl = document.createElement('div');
+      this.chartEl.className = 'vislib-chart';
+      this.container.appendChild(this.chartEl);
+
+      //this.el.className = 'vis-container';
     }
 
     render(esResponse) {
       this._response = esResponse;
       if (this.vis.vislibVis) {
         this.destroy();
-      } else {
-        this.vis.refreshLegend = 0;
       }
 
-      return new Promise((resolve) => {
+      return new Promise(async (resolve) => {
         if (this.el.clientWidth === 0 || this.el.clientHeight === 0) {
           return resolve();
         }
-        this.vis.vislibVis = new vislib.Vis(this.el, this.vis.params);
+
+        let $scope;
+        if (this.vis.params.addLegend) {
+          const legendPositionToVisContainerClassMap = {
+            top: 'vislib-container--legend-top',
+            bottom: 'vislib-container--legend-bottom',
+            left: 'vislib-container--legend-left',
+            right: 'vislib-container--legend-right',
+          };
+
+          const getVisContainerClasses = () => {
+            return legendPositionToVisContainerClassMap[this.vis.params.legendPosition];
+          };
+
+          // update the legend class on the parent element
+          $(this.container).attr('class', (i, cls) => {
+            return cls.replace(/vislib-container--legend-\S+/g, '');
+          }).addClass(getVisContainerClasses());
+
+          $scope = $rootScope.$new();
+          $scope.refreshLegend = 0;
+          $scope.vis = this.vis;
+          $scope.visData = esResponse;
+          $scope.uiState = $scope.vis.getUiState();
+          const legendHtml = $compile('<visualize-legend></visualize-legend>')($scope);
+          this.container.appendChild(legendHtml[0]);
+          $scope.$digest();
+          // We need to wait one digest cycle for the legend to render, before
+          // we want to render the chart, so it know about the legend size.
+          await new Promise(resolve => $timeout(resolve));
+        }
+
+        this.vis.vislibVis = new vislib.Vis(this.chartEl, this.vis.params);
         this.vis.vislibVis.on('brush', this.vis.API.events.brush);
         this.vis.vislibVis.on('click', this.vis.API.events.filter);
         this.vis.vislibVis.on('renderComplete', resolve);
         this.vis.vislibVis.render(esResponse, this.vis.getUiState());
+
+        if (this.vis.params.addLegend) {
```

Comment:
thats right. first we compile the legend template and add it to DOM, so the chart DOM element has the right size. Legend will show `loading` text at this time. then we call vislib render method, which will figure out what series we have in our data (this will happen synchronously) and then we refresh the legend (so it will show the right series).